File: Assets/Script/Example/Hello World.lua
-- [yue]: Script/Example/Hello World.yue
local Node = Dora.Node -- 1
local print = _G.print -- 1
local once = Dora.once -- 1
local sleep = Dora.sleep -- 1
local threadLoop = Dora.threadLoop -- 1
local App = Dora.App -- 1
local ImGui = Dora.ImGui -- 1
local Vec2 = Dora.Vec2 -- 1
do -- 3
	local _with_0 = Node() -- 3
	_with_0:slot("Enter", function() -- 4
		return print("on enter event") -- 4
	end) -- 4
	_with_0:slot("Exit", function() -- 5
		return print("on exit event") -- 5
	end) -- 5
	_with_0:slot("Cleanup", function() -- 6
		return print("on node destoyed event") -- 6
	end) -- 6
	_with_0:schedule(once(function() -- 7
		for i = 5, 1, -1 do -- 8
			print(i) -- 9
			sleep(1) -- 10
		end -- 10
		return print("Hello World!") -- 11
	end)) -- 7
end -- 3
local windowFlags = { -- 16
	"NoDecoration", -- 16
	"AlwaysAutoResize", -- 16
	"NoSavedSettings", -- 16
	"NoFocusOnAppearing", -- 16
	"NoNav", -- 16
	"NoMove" -- 16
} -- 16
return threadLoop(function() -- 24
	local width -- 25
	width = App.visualSize.width -- 25
	ImGui.SetNextWindowBgAlpha(0.35) -- 26
	ImGui.SetNextWindowPos(Vec2(width - 10, 10), "Always", Vec2(1, 0)) -- 27
	ImGui.SetNextWindowSize(Vec2(240, 0), "FirstUseEver") -- 28
	return ImGui.Begin("Hello World", windowFlags, function() -- 29
		ImGui.Text("Hello World (Yuescript)") -- 30
		ImGui.Separator() -- 31
		return ImGui.TextWrapped("Basic Dora schedule and signal function usage. Written in Yuescript. View outputs in log window!") -- 32
	end) -- 32
end) -- 32

File: Tools/dora-rust/dora-test/src/tests/hello_world.rs
/* Copyright (c) 2016-2026 Li Jin <dragon-fly@qq.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */

use dora_ssr::*;

pub fn test() {
	let mut node = Node::new();
	Slot::on_enter(&mut node, || {
		p!("on enter event");
	});
	Slot::on_exit(&mut node, || {
		p!("on exit event");
	});
	Slot::on_cleanup(&mut node, || {
		p!("on node destoyed event");
	});
	node.schedule(once(move |mut co| async move {
		for i in (1..=5).rev() {
			p!("{}", i);
			sleep!(co, 1.0);
		}
		p!("Hello World!");
	}));

	let mut imgui_node = Node::new();
	let windows_flags =
		ImGuiWindowFlag::NO_DECORATION |
		ImGuiWindowFlag::AlwaysAutoResize |
		ImGuiWindowFlag::NoSavedSettings |
		ImGuiWindowFlag::NoFocusOnAppearing |
		ImGuiWindowFlag::NO_NAV |
		ImGuiWindowFlag::NoMove;
	imgui_node.schedule(Box::new(move |_| {
		let width = App::get_visual_size().width;
		ImGui::set_next_window_bg_alpha(0.35);
		ImGui::set_next_window_pos_opts(&Vec2::new(width - 10.0, 10.0), ImGuiCond::Always, &Vec2::new(1.0, 0.0));
		ImGui::set_next_window_size_opts(&Vec2::new(240.0, 0.0), ImGuiCond::FirstUseEver);
		ImGui::begin_opts("Hello World", windows_flags, || {
			ImGui::text("Hello World (Rust)");
			ImGui::separator();
			ImGui::text_wrapped("Basic Dora schedule and signal function usage. View outputs in log window!");
		});
		false
	}));
}